### PR TITLE
Destination Bigquery+Snowflake: Fix resuming truncate refresh 

### LINKED ISSUE #41036: Destinations supporting refreshes: Resuming truncate refresh incorrectly skips T+D if the resumed sync emits no records
### Connector Name

destination-bigquery + destination-snowflake

### Connector Version

2.8.1

### What step the error happened?

During the sync

### Relevant information

example:
1. Attempt 1 of a truncate refresh emits 10 records, then crashes without emitting a stream COMPLETE status. We write these to a temp raw table, and skip T+D.
2. Attempt 2 emits 0 records. [This](https://github.com/airbytehq/airbyte/blob/0b0efbf073a5854155d10fa461a4ebe58334637f/airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt#L287) check will skip T+D.
3. We then overwrite the final table with an empty final table.

to fix: truncate syncs should also check the temp raw table initial status for whether there were preexisting records.

### Relevant log output

_No response_

### Contribute

- [X] Yes, I want to contribute

### PR DESCRIPTION
**Pull request by edgao**
closes https://github.com/airbytehq/airbyte/issues/41036

### CODE REVIEW

**Review by stephane-airbyte — APPROVED** [review]

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt [hunk 1]
```diff
@@ -27,8 +27,13 @@ abstract class AbstractStreamOperation<DestinationState : MinimumDestinationStat
...
     private val finalTmpTableSuffix: String
-    // null for truncate syncs, where we don't care at all about the initial status.
-    private val initialRawTableStatus: InitialRawTableStatus?
+    /**
+     * The status of the raw table that "matters" for this sync. Specifically:
+     * * For normal syncs / merge refreshes, this is the status of the real raw table)
+     * * For truncate refreshes, this is the status of the temp raw table (because we never even
+     * look at the real raw table)
+     */
+    private val initialRawTableStatus: InitialRawTableStatus
```
Comment (johnny-schmidt) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt:36]: @edgao  Remind me the difference here exactly. I really tried to follow this morning. :)

* normal sync/merge refresh: we populate a running raw table, then type-dedupe into the tmp table, then swap?
* for truncate refreshes: we populate a tmp raw table, then swap?

Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt:36]: for an overwrite, we populate a tmp raw table, truncate and dedupe into a tmp final table, and them swap both with the raw and final tables.
For an append, we don't do tmp tables.

truncate refresh is an overwrite. Everything else is an append

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt [hunk 2]
```diff
@@ -278,25 +283,32 @@ abstract class AbstractStreamOperation<DestinationState : MinimumDestinationStat
...
                 "Skipping typing and deduping for stream ${streamConfig.id.originalNamespace}.${streamConfig.id.originalName} " +
                     "because it had no records during this sync and no unprocessed records from a previous sync."
             }
-        } else if (isTruncateSync && (!streamSuccessful || syncSummary.recordsWritten == 0L)) {
+        } else if (
+            isTruncateSync &&
+                (!streamSuccessful ||
+                    (syncSummary.recordsWritten == 0L &&
+                        !(initialRawTableStatus.rawTableExists &&
+                            initialRawTableStatus.hasUnprocessedRecords)))
```
Comment (johnny-schmidt) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt:297]: Alright, I think I understand: before if we crashed and started again but had no more records, we'd incorrectly interpret that as there being no intermediate state. This clarifies: we need both 0 new records AND the tmp table not to exist.

**Review by johnny-schmidt — APPROVED** [review]

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt (hunk 2 above)
Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt:297]: yep. tried to capture that in the comment :/ but this is definitely an ugly conditional

File: docs/integrations/destinations/bigquery.md [hunk 3]
```diff
@@ -223,7 +223,8 @@ tutorials:
...
 |:--------|:-----------|:-----------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------|
-| 2.8.1   | 2024-06-25 | [39379](https://github.com/airbytehq/airbyte/pull/38331)   | Removing requirement of a redundant permission bigquery.datasets.create permission                                                                              |
+| 2.8.2   | 2024-07-08 | [41041](https://github.com/airbytehq/airbyte/pull/41041)   | Fix resume logic in truncate refreshes to prevent data loss                                                                                                     |
+| 2.8.1   | 2024-06-25 | [39379](https://github.com/airbytehq/airbyte/pull/39379)   | Removing requirement of a redundant permission bigquery.datasets.create permission                                                                              |
```
Comment (edgao) [inline docs/integrations/destinations/bigquery.md:227]: (fixing a copypasta error in the url - no other changes in this line)

**Pull request merged**